src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,164 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
+                if (bytesToCopy == 0)
+                {
+                    return -1;
+                }
+
+                var span = new Span<byte>(_byteArray, _offset, bytesToCopy);
+                span.CopyTo(output);
+                _offset += bytesToCopy;
+                _bytesAvailable -= bytesToCopy;
+                return bytesToCopy;
+            }
+
+            // Bio is already released by the ssl object
+            private void Dispose(bool isDisposing)
+            {
+                if (_handle.IsAllocated)
+                {
+                    _handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            ~ReadBioBuffer()
+            {
+                Dispose(false);
+            }
+        }
+
+        internal class WriteBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private byte[] _byteArray;
+            private int _bytesWritten;
+            private bool _isHandshake;
+
+            internal WriteBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.Crypto.ManagedSslBio.BioSetGCHandle(_bioHandle, _handle);
+            }
+
+            public int BytesWritten => _bytesWritten;
+
+            public void SetData(byte[] buffer, bool isHandshake)
+            {
+                Debug.Assert(_byteArray == null);
+
+                _byteArray = buffer;
+                _bytesWritten = 0;
+                _isHandshake = isHandshake;
+            }
+
+            public int TakeBytes(out byte[] output)
+            {
+                output = _byteArray;
+                var bytes = _bytesWritten;
+                Reset();
+                return bytes;
+            }
+
+            public void Reset()
+            {
+                _bytesWritten = 0;
+                _byteArray = null;
+            }
+
+            public int Write(Span<byte> input)
+            {
+                // Only for the handshake do we dynamically allocate
+                // buffers for normal encrypt operations we use a fixed
+                // size buffer handed to us and loop to do all the needed
+                // writes. This should be changed for the handshake as well
+                // but will require more securechannel/sslstatus changes
+                if (_isHandshake)
+                {
+                    if (_byteArray == null)
+                    {
+                        _byteArray = new byte[input.Length];
+                        _bytesWritten = 0;
+                    }
+
+                    else if (_byteArray.Length - _bytesWritten < input.Length)
+                    {
+                        var oldSpan = new Span<byte>(_byteArray);
```

Comment:
no reason other than you can just use span.copy which looks nicer than array copy. You want it changed?